File: 73.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Counting fractions in a
range". The problem is: How many fractions lie between 1/3 and 1/2 in the
sorted set of reduced proper fractions for d <= 12,000?
"""

from __future__ import division
import argparse


def main(args):
    """Counting fractions in a range"""

    # Constants
    LEFT_FRAC = [1, 3]
    RIGHT_FRAC = [1, 2]

    count = 0
    for d in range(2, args.limit + 1):
        start_n = int(LEFT_FRAC[0] / LEFT_FRAC[1] * d)
        end_n = int(RIGHT_FRAC[0] / RIGHT_FRAC[1] * d) + 1
        for n in range(start_n, end_n + 1):
            if compare_fractions([n, d], LEFT_FRAC) <= 0:
                continue
            if compare_fractions([n, d], RIGHT_FRAC) >= 0:
                break
            if gcd(n, d) == 1:
                count += 1

    print(count)


def compare_fractions(frac1, frac2):
    """Compare two fractions"""

    big_numer1 = frac1[0] * frac2[1]
    big_numer2 = frac2[0] * frac1[1]

    if big_numer1 > big_numer2:
        return(1)
    elif big_numer1 < big_numer2:
        return(-1)
    else:
        return(0)


def gcd(a, b):
    """Get greatest common divisor"""

    if a > b:
        a, b = b, a

    while a:
        a, b = b % a, a

    return(b)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Counting fractions in a range')
    parser.add_argument(
        'limit', metavar='LIMIT', type=int, default=12000, nargs='?',
        help='The limit on d')
    args = parser.parse_args()

    main(args)

File: 73.R
# Counting fractions in a range

Args  <- commandArgs()
limit <- ifelse(is.na(Args[6]), 12000, as.numeric(Args[6]))

# Constants
LEFT_FRAC <- c(1, 3)
RIGHT_FRAC <- c(1, 2)

compare_fractions <- function(frac1, frac2) {
    big_numer1 = as.numeric(frac1[1]) * as.numeric(frac2[2])
    big_numer2 = as.numeric(frac2[1]) * as.numeric(frac1[2])

    if ( big_numer1 > big_numer2 ) {
        return(1)
    } else if ( big_numer1 < big_numer2 ) {
        return(-1)
    } else {
        return(0)
    }
}

# Get greatest common divisor
gcd <- function(a, b) {
    if ( a > b ) {
        a.tmp <- a
        a <- b
        b <- a.tmp
    }

    while ( a ) {
        a.tmp <- a
        a <- b %% a
        b <- a.tmp
    }

    return(b)
}

count <- 0
for ( d in seq(2, limit) ) {
    start_n <- floor(LEFT_FRAC[1] / LEFT_FRAC[2] * d)
    end_n <- floor(RIGHT_FRAC[1] / RIGHT_FRAC[2] * d) + 1
    for ( n in seq(start_n, end_n) ) {
        if ( compare_fractions(c(n, d), LEFT_FRAC) <= 0 ) {
            next
        }
        if ( compare_fractions(c(n, d), RIGHT_FRAC) >= 0 ) {
            break
        }
        if ( gcd(n, d) == 1 ) {
            count <- count + 1
        }
    }
}

cat(count, fill=TRUE)
